### DISCUSSION TITLE
Custom Labels for kubernetes_executor

### DISCUSSION THREAD

**Discussion by kl4072** [apache/airflow#37553] [Q&A]
we have some corporate policy that doesnt allow us to create dynamic worker pods without specific labels. 

On the kubernetes executor is there a way to add custom labels? when deploying via helm i have tried all the label options.

under kubernetes_executor can we add custom labels?

Running Airflow 2.8.1 version

{rest.py:231} DEBUG - response body: {"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"admission webhook \"va.sh\" denied the request: [allow-labels-only] All pods must have labels of owner and appName. ","reason":"Forbidden","code":403}

[2024-02-19T20:06:10.553+0000] {kubernetes_executor_utils.py:323} ERROR - Exception when attempting to create Namespaced Pod: {

**Comment by potiuk** [comment]
Did you try custom pod template? https://airflow.apache.org/docs/apache-airflow/stable/core-concepts/executor/kubernetes.html#pod-template-file  ? Can you report here if it solved your problem when you experiment with it?

**Comment by kl4072** [comment]
I have tried the following podTemplate : 'files/pod-template.yaml'  values.yaml

podTemplate: |
  apiVersion: v1
  kind: Pod
  metadata:
    name: placeholder-name
    labels:
      tier: airflow
      component: worker
      appName: airflow-webserver
      owner: gdp
  spec:
    containers:
      - env:
          - name: AIRFLOW__CORE__EXECUTOR
            value: KubernetesExecutor
          - name: AIRFLOW__CORE__FERNET_KEY
            value: "test1"
          - name: AIRFLOW__DATABASE__SQL_ALCHEMY_CONN
            value: postgresql+psycopg2://..:5432/airflow
          - name: AIRFLOW_CONN_AIRFLOW_DB
            value: postgresql+psycopg2://...:5432/airflow
        image: airflow/airflow:2.8.1-python3.11-1
        imagePullPolicy: IfNotPresent
        name: base
        volumeMounts:
          - mountPath: "/opt/airflow/logs"
            name: airflow-logs
          - mountPath: /opt/airflow/dags
            name: airflow-dags
            readOnly: true
          - mountPath: /opt/airflow/airflow.cfg
            name: airflow-config
            readOnly: true
            subPath: airflow.cfg
    restartPolicy: Never
    securityContext:
      runAsUser: 50000
      fsGroup: 50000
    serviceAccountName: airflow-service-account
    volumes:
      - name: airflow-dags
        persistentVolumeClaim:
          claimName: airflow-dags
      - emptyDir: {}
        name: airflow-logs
      - configMap:
          name: airflow-airflow-config
        name: airflow-config

this didnt seem to change the pod template for worker pod

**Comment by kl4072** [comment]
i got past the error after mounting the pod_template_yaml and adding the labels there      
`volumeMounts:
            - name: config
              mountPath: {{ include "airflow_pod_template_file" . }}/pod_template_file.yaml
              subPath: pod_template_file.yaml
              readOnly: true`

**Comment by kl4072** [comment]
There is a new issue however, since i cannot create new Role's in the namespace i am using a Role created and bound by the admin so i have disabled the creation of pod-launcher-role, rolebinding  .. does the Role need to list resources in default namespace? 

HTTP response body: b'{\n  "kind": "Status",\n  "apiVersion": "v1",\n  "metadata": {},\n  "status": "Failure",\n  "message": "pods is forbidden: User \\"system:serviceaccount:orchestration-dev:airflow-service-account\\" cannot list resource \\"pods\\" in API group \\"\\" in the namespace \\"default\\"",\n  "reason": "Forbidden",\n  "details": {\n    "kind": "pods"\n  },\n  "code": 403\n}'

**Reply by potiuk** [reply]
No idea. I suggest to start new thread for new issue, then you get yourself a chance.to get answers from people that will decide they can help in the other issue

**Reply by kl4072** [reply]
Thanks! will create a new question
